### PR TITLE
gh-111062: Reusable Windows build that supports free-threaded mode as the conditional CI

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-111062

### CODE REVIEW

**Comment by corona10** [comment]
co-work with @jota12x

File: .github/workflows/reusable-build-win.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  workflow_call:
+    inputs:
+      free-threaded:
+        required: false
+        type: boolean
+        default: false
+
+jobs:
+  build_win_32:
+    name: 'build and test (x86)'
```
Comment (corona10) [inline .github/workflows/reusable-build-win.yml]: @hugovk I would like to use the name as `build and test` variations for macOS and Linux too.
It looks more clear when we look at the CI status bar.
But I will do it as a separate PR.

File: .github/workflows/reusable-build-win.yml
Comment (hugovk) [inline .github/workflows/reusable-build-win.yml:1]: Let's include the fullname `windows` in this filename.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -178,56 +178,19 @@ jobs:
...
-    - uses: actions/checkout@v4
-    - name: Register MSVC problem matcher
-      run: echo "::add-matcher::.github/problem-matchers/msvc.json"
-    - name: Build CPython
-      run: .\PCbuild\build.bat -e -d -p x64
-    - name: Display build info
-      run: .\python.bat -m test.pythoninfo
-    - name: Tests
-      run: .\PCbuild\rt.bat -p x64 -d -q --fast-ci
+    uses: ./.github/workflows/reusable-build-win.yml
```
Comment (hugovk) [inline .github/workflows/build.yml]:
```suggestion
    uses: ./.github/workflows/reusable-build-windows.yml
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -178,56 +178,19 @@ jobs:
...
-    env:
-       IncludeUwp: 'true'
-    steps:
-    - uses: actions/checkout@v4
-    - name: Register MSVC problem matcher
-      run: echo "::add-matcher::.github/problem-matchers/msvc.json"
-    - name: Build CPython
-      run: .\PCbuild\build.bat -e -d -p arm64
+    if: needs.check_source.outputs.run_tests == 'true' && contains(github.event.pull_request.labels.*.name, 'topic-free-threaded')
+    uses: ./.github/workflows/reusable-build-win.yml
```
Comment (hugovk) [inline .github/workflows/build.yml]:
```suggestion
    uses: ./.github/workflows/reusable-build-windows.yml
```

**Comment by corona10** [comment]
@hugovk  For https://github.com/python/cpython/pull/110764, I merged the PR. Let's do it at the separate PR if there is something what we should do

**Comment by vstinner** [comment]
On a doc PR, the CI failed because Windows was skipped: https://github.com/python/cpython/actions/runs/6696696148/job/18198167333?pr=111512

It's surprising that once the job is called `build_windows`, but later `build_win_amd64` is allowed to be skipped. Is it a typo?

cc @Eclips4

**Comment by vstinner** [comment]
Ah, https://github.com/python/cpython/pull/111521 was proposed to fix the job.

**Pull request merged**